rust/main/chains/hyperlane-sovereign/src/provider/rest_client.rs

```diff
@@ -1,179 +1,1127 @@
+            .http_get(&query)
+            .await
+            .map_err(|e| ChainCommunicationError::CustomError(format!("HTTP Get Error: {e}")))?;
+        let response: Schema<Tx> = serde_json::from_slice(&response)?;
+        let data = response.data.ok_or(ChainCommunicationError::CustomError(
+            "Invalid response".to_string(),
+        ))?;
+        Ok(data)
     }
 
-    pub async fn get_chain_metrics(&self) -> ChainResult<Option<ChainInfo>> {
-        todo!()
+    // Return the latest slot, and the highest committed batch number in that slot.
+    pub async fn get_latest_slot(&self) -> ChainResult<(u32, Option<u32>)> {
+        info!("get_latest_slot(&self)");
+        #[derive(Clone, Debug, Deserialize)]
+        struct BatchRange {
+            // start: u32,
+            end: u32,
+        }
+        #[derive(Clone, Debug, Deserialize)]
+        struct Data {
+            batch_range: BatchRange,
+            number: u32,
+        }
+
+        let query = "/ledger/slots/latest?children=0";
+        let response = self
+            .http_get(query)
+            .await
+            .map_err(|e| ChainCommunicationError::CustomError(format!("HTTP Get Error: {e}")))?;
+        let response: Schema<Data> = serde_json::from_slice(&response)?;
+        let data = response.data.ok_or(ChainCommunicationError::CustomError(
+            "Invalid response".to_string(),
+        ))?;
+
+        // bach_range.end is exclusive - it's one above the last committed batch number
+        let last_batch = if data.batch_range.end > 0 {
+            Some(data.batch_range.end - 1)
+        } else {
+            None
+        };
+
+        Ok((data.number, last_batch))
     }
 
-    pub async fn get_count(&self, lag: Option<NonZeroU64>) -> ChainResult<u32> {
-        let query = match lag {
-            Some(lag) => format!("/modules/mailbox/state/nonce?rollup_height={}", lag),
+    // @Provider - test working, need to test all variants
+    pub async fn is_contract(&self, key: H256) -> ChainResult<bool> {
+        info!("is_contract(&self, key: &str) key:{:?}", key);
+        #[derive(Clone, Debug, Deserialize)]
+        struct Data {
+            key: Option<String>,
+            _value: Option<String>,
+        }
+
+        for module in [
+            "mailbox-hook-registry",
+            "mailbox-ism-registry",
+            "mailbox-recipient-registry",
+        ] {
+            let query = format!(
+                "/modules/{}/state/registry/items/{}",
+                module,
+                to_bech32(key)?
+            );
+
+            let response = self.http_get(&query).await.map_err(|e| {
+                ChainCommunicationError::CustomError(format!("HTTP Get Error: {e}"))
+            })?;
+            let response: Schema<Data> = serde_json::from_slice(&response)?;
+            println!("{response:?}");
+
+            if let Some(data) = response.data {
+                return Ok(data.key.is_some());
+            }
```

Comment:
We can avoid unnecessary intermediate variables.

```suggestion
            if response.data.and_then(|data| data.key).is_some() {
                return Ok(true);
            }
```